File: src/main/java/discretemath/utils/CacheValue.java
package discretemath.utils;

import java.util.NoSuchElementException;
import java.util.concurrent.atomic.AtomicReference;

public class CacheValue<T> {

	private final AtomicReference<T> value = new AtomicReference<>();

	public boolean isSet() {
		return value.get() != null;
	}

	public T get() {
		T value = this.value.get();
		if (value == null) {
			throw new NoSuchElementException();
		}

		return value;
	}

	public void set(T value) {
		boolean changed = this.value.compareAndSet(null, value);

		if (!changed) {
			throw new IllegalStateException();
		}
	}
}

File: src/main/scala/discretemath/utils/CacheValue.scala
package discretemath.utils

import java.util.NoSuchElementException
import java.util.concurrent.atomic.AtomicReference

class CacheValue[T] {
	private val value = new AtomicReference[T]

	def isSet: Boolean = value.get != null

	def get: T = {
		val value = this.value.get
		if (value == null) throw new NoSuchElementException
		value
	}

	def set(value: T): Unit = {
		val changed = this.value.compareAndSet(null.asInstanceOf[T], value)
		if (!changed) throw new IllegalStateException
	}
}